### DISCUSSION TITLE
UR5 Servoing

### DISCUSSION THREAD

**Discussion by adnan-saood** [UniversalRobots/Universal_Robots_ROS2_Driver#325] [Q&A] [answered]
Greetings all.
I am trying to control a UR5 robot using ROS2. The goal is to follow a polygon path with end-effector velocity control. The velocity of the end-effector needs to be controlled precisely (the task is similar to painting or welding but on a very small scale (area 2cmx2cm).

I started by using the moveit2 move group interface in C++, the problem is there seems to be no way to control the velocity of the end-effector, only timestamp it, and the results are not smooth.

I then started using moveit2-servo. I wrote a code for pose tracking, where the end-effector follows a frame published on a topic. The ros2_controller was (JointTrajectoryController). I was able to control the velocity and position of the end-effector using this method in simulation, but when I tried it on the real UR5 robot, the robot experienced rough vibrations. I understand that this behavior is expected from the JointTrajectoryController controller.

I tried to control the robot using the jointGroupVelocityController [(link)](https://github.com/ros-controls/ros2_controllers/tree/master/velocity_controllers), but with no result. I am a bit confused if this is at all possible using the UR-ROBOTS-ROS2-Driver, as I found the plugin definition for the position, but without a mention of velocity interface. [(link)](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/ur_robot_driver/src/hardware_interface.cpp)

I feel I am missing something. Any suggestions would be appreciated. Thank you for your time.

**Comment by fmauch** [comment]
What do you mean with not mentioning the velocity interface? How about [here](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/ur_robot_driver/src/hardware_interface.cpp#L89) or [here](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/ff3517dff7bd69fafd755b07b6fecc31b0a3ebc5/ur_robot_driver/src/hardware_interface.cpp#L136) just to pick two examples.

Did you try to change the controller using `ros2 control switch_controllers --start forward_velocity_controller --stop scaled_joint_trajectory_controller`?

**Comment by adnan-saood** [comment, marked as answer]
Hello,
Thanks for your reply and for the answer [here](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/issues/327#issuecomment-1091695669), I was able to use moveit_servo with the UR5 robot via the URSim simulator.
